File: sim/agent/police.ts
import { Simulator } from 'simulator'
import { Agent }     from './agent'
import { Behavior }  from './behavior'

export class PoliceBehavior implements Behavior {

  constructor(private sim: Simulator) {}

  init(agent: Agent): void {
    // Randomly choose a starting location
    agent.location = this.sim.rng.pick(this.sim.arena.nodes)
  }

  move(agent: Agent): void {
    // Police agents move randomly
    const current = agent.location

    // Pick a random path to walk down
    const edge = this.sim.rng.pick(current.edges)
    edge.follow(agent)

    // console.log(`${agent.id}: move ${current.id} -> ${agent.location.id}`)
  }

  action(_: Agent): void {}

}

File: sim/police.go
package sim

// NewPoliceAgent creates a new Police agent.
func NewPoliceAgent(id uint64) *Agent {
	return &Agent{
		ID:       id,
		Kind:     PoliceAgentKind,
		Behavior: []Behavior{&PoliceBehavior{}},
	}
}

// PoliceBehavior picks a random starting location, and moves about randomly.
type PoliceBehavior struct{ baseBehavior }

// Init causes the police agent to pick a random starting location.
func (police *PoliceBehavior) Init(p *Provider, agent *Agent) {
	// Pick a random node.
	node := p.RNG().Node(p.Arena().Nodes)

	// Set it as the starting location
	node.Enter(agent)

	node.TotalPolice++
}

// DayStart is run on the first tick of each simulation day.
func (police *PoliceBehavior) DayStart(p *Provider, agent *Agent) {}

// Move casues the police agent to pick a random edge at its current location
// and travel down it.
func (police *PoliceBehavior) Move(p *Provider, agent *Agent) {
	// Pick a random edge.
	edge := p.RNG().Edge(agent.Location.Edges)

	// Walk down the edge
	edge.Follow(agent)

	agent.TravelDistance++
	agent.Location.TotalPolice++
}
